payments/models.py

```diff
@@ -0,0 +1,369 @@
+from datetime import datetime, timedelta
+from decimal import Decimal
+
+from django.conf import settings
+from django.core.exceptions import ValidationError
+from django.core.validators import MinValueValidator
+from django.db import models
+from django.db.models import OuterRef, Q, Subquery
+from django.utils import translation
+from django.utils.formats import localize
+from django.utils.functional import cached_property
+from django.utils.timezone import now, utc
+from django.utils.translation import ugettext_lazy as _
+from rest_framework import serializers
+
+from resources.models import Reservation, Resource
+from resources.models.utils import generate_id
+
+from .exceptions import OrderStateTransitionError
+from .utils import convert_aftertax_to_pretax, get_price_period_display, rounded
+
+# The best way for representing non existing archived_at would be using None for it,
+# but that would not work with the unique_together constraint, which brings many
+# benefits, so we use this sentinel value instead of None.
+ARCHIVED_AT_NONE = datetime(9999, 12, 31, tzinfo=utc)
+
+TAX_PERCENTAGES = [Decimal(x) for x in (
+    '0.00',
+    '10.00',
+    '14.00',
+    '24.00',
+)]
+
+DEFAULT_TAX_PERCENTAGE = Decimal('24.00')
+
+
+class ProductQuerySet(models.QuerySet):
+    def current(self):
+        return self.filter(archived_at=ARCHIVED_AT_NONE)
+
+    def rents(self):
+        return self.filter(type=Product.RENT)
+
+
+class Product(models.Model):
+    RENT = 'rent'
+    EXTRA = 'extra'
+    TYPE_CHOICES = (
+        (RENT, _('rent')),
+        (EXTRA, _('extra')),
+    )
+
+    PRICE_PER_PERIOD = 'per_period'
+    PRICE_FIXED = 'fixed'
+    PRICE_TYPE_CHOICES = (
+        (PRICE_PER_PERIOD, _('per period')),
+        (PRICE_FIXED, _('fixed')),
+    )
+
+    created_at = models.DateTimeField(verbose_name=_('created at'), auto_now_add=True)
+
+    # This ID is common to all versions of the same product, and is the one
+    # used as ID in the API.
+    product_id = models.CharField(max_length=100, verbose_name=_('internal product ID'), editable=False, db_index=True)
+
+    # archived_at determines when this version of the product has been either (soft)
+    # deleted or replaced by a newer version. Value ARCHIVED_AT_NONE means this is the
+    # current version in use.
+    archived_at = models.DateTimeField(
+        verbose_name=_('archived_at'), db_index=True, editable=False, default=ARCHIVED_AT_NONE
+    )
+
+    type = models.CharField(max_length=32, verbose_name=_('type'), choices=TYPE_CHOICES, default=RENT)
+    sku = models.CharField(max_length=255, verbose_name=_('SKU'))
+    name = models.CharField(max_length=100, verbose_name=_('name'), blank=True)
+    description = models.TextField(verbose_name=_('description'), blank=True)
+
+    price = models.DecimalField(
```

Comment:
When dealing with monetary values, it can be easier to save the amount in the smallest unit (cent, penny etc) to avoid all kinds of hurdles with rounding etc. and just convert it to decimal representation for viewing. There are even some currencies in the world which do not have fractional units (yen IIRC) where decimals are not used.

Definitely no reason to change anything, just a random comment :)

Reply:
Not sure how that would actually work, as more complex tax / order / order line price calculations can have pretty specific rules on when and how to do rounding, and even with these most basic calculations that we have we still need to round 5 up.

Reply:
Probably true. Some cases would be simpler with just integer arithmetic, but I guess there's still the problem of rounding to the cent.